### ISSUE TITLE
[ENHANCEMENT] Update (unit) testing and GitHub continuous integration with tox

### ISSUE THREAD

**Issue by martinvonk** [pastas/pastas#509]
**Describe the proposed enhancement**
This enhancement proposes two a few things:
- improve testing by making tests more specific by separate testing for python versions, formatting etc.
- allow ALL tests to run locally in different (Python) environments

I want to do a few things:

1.  Improve the `pytest` implementation using fixtures for models.
Currently our tests have a lot of repetition test with a new model being created for every test function:
```
def test_add_stressmodel():
    ml = ps.Model(obs, name="Test_Model")
    sm = ps.RechargeModel(prec=rain, evap=evap, rfunc=ps.Gamma(), name="rch")
    ml.add_stressmodel(sm)
    return None

def test_add_stressmodels():
    ml = ps.Model(obs, name="Test_Model")
    sm1 = ps.StressModel(rain, rfunc=ps.Exponential(), name="prec")
    sm2 = ps.StressModel(evap, rfunc=ps.Exponential(), name="evap")
    ml.add_stressmodel([sm1, sm2])
    return None
```

However, this could be simplified (and sped up?) by using pytest fixtures.

```
@pytest.fixture
def ml():
    return ps.Model(obs, name="Test_Model")

def test_add_stressmodel(ml):
    sm = ps.RechargeModel(prec=rain, evap=evap, rfunc=ps.Gamma(), name="rch")
    ml.add_stressmodel(sm)
    return None

```

2. Use [tox](https://tox.wiki/en/latest/) to automate and standardize the testing. tox is an environment orchestrator. Use it to define how to setup and execute various tools on your projects. This can all be done locally so we can test it before committing to GitHub. Additionally we can define different routines for all the tests we want to run, e.g.:

- test runners (such as pytest and hopefully mypy in the future #480)
- linters (flake8)
- formatters (black and isort)
- documentation generators (sphinx)

This also makes it more clear which tests fail on GitHub ci, so we won't just see python 3.11 failed because it failed on black. But you see the formatting test failed.

**Comment by martinvonk** [comment]
3. Test notebooks from documentation when releasing a new version on master.

**Comment by dbrakenhoff** [comment]
Forgot to post my comment here :innocent: :

I address a small part of this in #515. Separating linting and testing.

**Comment by raoulcollenteur** [comment]
Looks like all this is done in PR #515 now?

**Comment by martinvonk** [comment]
> Looks like all this is done in PR #515 now?

No, we only need to run the benchmark notebooks when building RTD on pull requests to the master branch (point 3.)

**Comment by raoulcollenteur** [comment]
Good point. I'd be happy if we could implement this with 1.0. Should not be too hard now right?

**Comment by martinvonk** [comment]
> Good point. I'd be happy if we could implement this with 1.0. Should not be too hard now right?

No probably not but I don't have time for this today so will probably be 1.1 (or 1.0.1 :))

**Comment by martinvonk** [comment]
With #583 should be good now. Small bug fixed in #588 (running benchmark notebooks with normal unit tests).
